ocaml/typing/typecore.ml

```diff
@@ -854,14 +854,24 @@ let has_poly_constraint spat =
 
 (** Mode cross a left mode *)
 let mode_cross_left env ty mode =
-  let mode =
-    if not (is_principal ty) then mode else
+  if not (is_principal ty) then
+    Value.disallow_right mode
+  else begin
     let jkind = type_jkind_purely env ty in
     let upper_bounds = Jkind.get_modal_upper_bounds jkind in
+    let upper_bounds =
+      Alloc.Const.merge
```

Comment:
I don't love this step. Maybe it makes sense to have a `meet_comonadic_subtract_monadic` function that does the right thing?

Reply:
I agree this ugly, but it disappears in my next PR, so probably best to leave it for now.